lib/ecstatic.js

```diff
@@ -245,6 +241,39 @@ var ecstatic = module.exports = function (dir, options) {
         status['500'](res, next, { error: err });
       });
     }
+
+    // Do a strong or weak etag comparison based on setting
+    // https://www.ietf.org/rfc/rfc2616.txt Section 13.3.3
+    function shouldReturn304(req, stat) {
+      if (!req || !req.headers) {
+        return false;
+      }
+
+      var modifiedSince = req.headers['if-modified-since'],
+        clientEtag = req.headers['if-none-match'],
+        serverEtag = etag(stat, opts.weakEtags);
+
+      if (!modifiedSince && !clientEtag) {
+        // Client did not provide any conditional caching headers
+        return false;
+      }
+
+      // If any of the headers provided don't match, then don't return 304
+      if (modifiedSince && (new Date(Date.parse(modifiedSince))) < stat.mtime) {
+        return false;
+      }
+
+      if (clientEtag) {
```

Comment:
I believe this check is still required. At this point, there's either no modifiedSince header OR a modifiedSince header that falls within the proper date range. We don't know that the clientEtag header exists for sure.